src/yetibot/models/postal_code.clj

```diff
@@ -0,0 +1,52 @@
+(ns yetibot.models.postal-code
+  (:require
+   [clojure.string :as str]))
+
+;;
+;; Postal Code validation, qualification, and cleanup
+;;
+
+(defn- nl-cleanup
+  [d s]
+  (str d " " (str/upper-case s)))
+
+(defn gb-cleanup
+  [a b c d]
+  (->> (filter #(not (nil? %)) [a b c d])
+       (str/join " ")
+       str/upper-case))
+
+(def postal-codes
+  {"US" {:re #"(?x) (\d{5}) (?: [-+] \d{4} )?"}
+
+   "RO" {:re #"(\d{6})"}
+
+   "BR" {:re #"(\d{5}-\d{3})"}
+
+   ;; https://en.wikipedia.org/wiki/Postal_codes_in_the_Netherlands
+   "NL" {:re #"(?ix) (\d{4}) \s* ([a-rt-z][a-z] | s[bce-rt-z])"
+         :cleanup nl-cleanup}
+   
+   ;; https://en.wikipedia.org/wiki/Postcodes_in_the_United_Kingdom#Validation
+   "GB" {:re #"(?ix) (?: ([a-z][a-hj-y]?[0-9][a-z0-9]?) \s* ([0-9][a-z]{2}) |
+                         (gir) \s* (0a{2}) )"
+         :cleanup gb-cleanup}
+   
+   "AU" {:re #"(\d{4})"}
+
+   "PH" {:re #"(\d{4})"}})
+
+(defn- pc-chk-clean
+  [s postal-codes]
+  (reduce-kv (fn [_ cc {:keys [:re :cleanup]}]
+            (when-let [[_ & groups] (re-matches re s)]
+              (let [cleanup (or cleanup (partial str))]
+                (reduced [cc (apply cleanup groups)]))))
+          nil
+          postal-codes))
+
```

Comment:
@devth - using reduce here after some of our prior discussions. This is a slightly different way of using reduce than I have in the past - I'm not reducing an input collection, but rather using reduce as an iterator, breaking out with reduced or returning nil if I fail to find a match. This feels a little backwards and I just wanted to get your take on it. What would be the most idiomatic approach here?

Reply:
Interesting usage of `reduced`! But yeah doesn't really feel like idiomatic Clojure. I don't think I've ever used it. On the plus side it's probably the most performant since you avoid multiple `re-matches`.

If aiming for readability and idiomatic code I'd probably split out a separate tiny function that helps find the matching postal-code entry like:

```clojure
(defn- find-postal-code
  "Attempt to find matching postal code CC and corresponding regex and optional cleaner"
  [s postal-codes]
  (first (filter (fn [[_ {:keys [re]}]] (re-matches re s)) postal-codes)))
```

Easy to read, does one thing, easy to test, etc.

Then:

```clojure
  (when-let [[cc {:keys [re cleanup]}] (find-postal-code s postal-codes)]
    (let [[_ & groups] (re-matches re s)
          cleanup (or cleanup (partial str))]
      [cc (apply cleanup groups)]))
```

Reply:
Playing with this now. Thanks!